### ISSUE TITLE
How to change between train and evaluate mode in forward method?

### ISSUE THREAD

**Issue by polars05** [PyTorchLightning/pytorch-lightning#1137]
I'm trying to use FasterRCNN from `torchvision.models.detection` which behaves in the following manner:
-  During training, the model expects both the input tensors, as well as targets and will return a dict containing the classification and regression losses:
```
loss_dict = model(images, targets)
```
- During evaluation, the model requires only the input tensors, and returns the post-processed
    predictions:
```
predictions = model(images)
```
How can I toggle between these two modes in `def forward(self, images):` of a PyTorch Lightning system?

**Comment by awaelchli** [comment]
How about this:
```
def forward(self, images, targets=None):
    if targets is not None:
        # Code for training
        output = ...
    else:
        # Code for prediction
        output = ...
    return output
```
In the if/else clause you could also dispatch to e.g. `self.predict(...)` and e.g. `self.forward_train(...)` to further break down the complexity of the forward pass.

**Comment by Borda** [comment]
@awaelchli thx for help! @polars05 feel free to re-open if needed... :]

**Comment by jzazo** [comment]
Can I trust model flags: `model.training` and `model.testing` to determine fit stage in forward method?

I am using an unsupervised method, so I don't have labels to determine the stage. I still need to know what stage the model is in to perform masking to the input, for example, and calculate loss.

Thank you!
